File: backend/fretquiz/src/main/java/com/kdp/fretquiz/game/Game.java
package com.kdp.fretquiz.game;

import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import com.kdp.fretquiz.game.ImmutableGame;
import com.kdp.fretquiz.user.User;
import org.immutables.value.Value;

import java.util.List;
import java.util.Set;

@Value.Immutable
@JsonSerialize(as = ImmutableGame.class)
public abstract class Game
{
    public abstract Long id();
    public abstract Status status();
    public abstract Opts opts();
    public abstract Long hostId();
    public abstract Set<User> users();
    public abstract List<Round> rounds();

    public enum Status
    {
        INIT,
        PLAYING,
        ROUND_OVER,
        GAME_OVER
    }
}

File: frontend/fretquiz/src/game/Game.ts
import { User } from "../user/user";

export type Status = 'INIT' | 'PLAYING' | 'ROUND_OVER' | 'GAME_OVER';

export interface Fretboard {
  tuning: string[];
  startFret: number;
  endFret: number;
}

export interface Opts {
  roundCount: number;
  fretboard: Fretboard;
  strings: number[];
  accidentals: string[]
}

export interface Guess {
  userId: number;
  isCorrect: boolean;
}

export interface Round {
  noteToGuess: string;
  guesses: Guess[];
  isOver: boolean;
}

export interface Game {
  id: number;
  status: Status;
  opts: Opts;
  hostId: number;
  users: User[];
  rounds: Round[];
}
